src/Lean/Data/Trie.lean

```diff
@@ -1,110 +1,202 @@
+  where
+    go : Trie α → StateM (Array α) Unit
+      | leaf a? => do
+        if let some a := a? then
+          modify (·.push a)
+      | node1 a? _ t' => do
+        if let some a := a? then
+          modify (·.push a)
+        go t'
+      | node a? _ ts => do
+        if let some a := a? then
+          modify (·.push a)
+        ts.forM fun t' => go t'
+
+/-- Returns all values whose key have the given string `pre` as a prefix, in no particular order. -/
+partial def findPrefix (t : Trie α) (pre : String) : Array α := go t 0
+  where
+    go (t : Trie α) (i : Nat) : Array α :=
+      if h : i < pre.utf8ByteSize then
+        let c := pre.getUtf8Byte i h
+        match t with
+        | leaf _val => .empty
+        | node1 _val c' t' =>
+          if c == c'
+          then go t' (i + 1)
+          else .empty
+        | node _val cs ts =>
+          match cs.findIdx? (· == c) with
+          | none   => .empty
+          | some idx => go (ts.get! idx) (i + 1)
+      else
+        t.values
+
+/-- Find the longest _key_ in the trie that is contained in the given string `s` at position `i`,
+and return the associated value. -/
+partial def matchPrefix (s : String) (t : Trie α) (i : String.Pos) : Option α :=
   let rec loop
-    | Trie.Node v m, i, acc =>
-      match s.atEnd i with
-      | true  => updtAcc v i acc
-      | false =>
-        let acc := updtAcc v i acc
-        let c   := s.get i
-        let i   := s.next i
-        match RBNode.find compare m c with
-        | some t => loop t i acc
-        | none   => acc
-  loop t i (i, none)
+    | leaf v, _, res =>
+      if v.isSome then v else res
+    | node1 v c' t', i, res =>
+      let res := if v.isSome then v else res
+      if h : i < s.utf8ByteSize then
+        let c := s.getUtf8Byte i h
+        if c == c'
+        then loop t' (i + 1) res
+        else res
+      else
+        res
+    | node v cs ts, i, res =>
+      let res := if v.isSome then v else res
+      if h : i < s.utf8ByteSize then
+        let c := s.getUtf8Byte i h
+        match cs.findIdx? (· == c) with
+        | none => res
+        | some idx => loop (ts.get! idx) (i + 1) res
+      else
+        res
+  loop t i.byteIdx none
 
 private partial def toStringAux {α : Type} : Trie α → List Format
-  | Trie.Node _ map => map.fold (fun Fs c t =>
-   format (repr c) :: (Format.group $ Format.nest 2 $ flip Format.joinSep Format.line $ toStringAux t) :: Fs) []
+  | leaf _ => []
+  | node1 _ c t =>
+    [ format (repr c), Format.group $ Format.nest 4 $ flip Format.joinSep Format.line $ toStringAux t ]
+  | node _ cs ts =>
+    List.join $ List.zipWith (fun c t =>
+      [ format (repr c), (Format.group $ Format.nest 4 $ flip Format.joinSep Format.line $ toStringAux t) ]
+    ) cs.toList ts.toList
 
 instance {α : Type} : ToString (Trie α) :=
   ⟨fun t => (flip Format.joinSep Format.line $ toStringAux t).pretty⟩
+
 end Trie
 
-end Parser
-end Lean
+end Data
+end Lean
```

Comment:
Our VS Code settings should prevent the final newline from being stripped, are you using a different editor?

Reply:
I am using VS Code, but am no expert in it. Maybe because I used `code src/` rather than `code .`, (wrongly?) assuming I have to do that to benefit from the `elan` override in `src/` that’s described in `doc/dev/index.md`.

Reply:
Good question! I believe you are right /cc @mhuisi

Reply:
VS Code will use the `.vscode/settings.json` of the workspace that you open. If you opened `src`, then it will not use it.

I'll see if VS Code can be more helpful when opening the wrong folder in Lean 4 itself, too.